### PR TITLE
[Experimental] Drop fast faces

### PR DESCRIPTION
**Pull request by numberZero**
@Fixer-007 @paramat Can you test performance, please? This mostly affects CPU, and only solid and liquid nodes.
I plan to base several features on this part, but if it affects performance significantly I will have to rethink my decision.

### CODE REVIEW

**Comment by numberZero** [comment]
Note: as a side effect, this fixes #6777 .
Note: Travis job errored, that’s its fault, not mine.

File: src/content_mapblock.cpp [hunk 1]
```diff
@@ -373,12 +379,46 @@ void MapblockMeshGenerator::drawAutoLightedCuboid(aabb3f box, const f32 *txc,
...
 	} else {
-		drawCuboid(box, tiles, tile_count, nullptr, txc);
+		drawCuboid(box, tiles, tile_count, nullptr, txc, faces);
 	}
 }
 
+void MapblockMeshGenerator::drawSolidNode()
+{
+	bool faces[6] = {0, 0, 0, 0, 0, 0};
+	static const v3s16 tile_dirs[6] = {
```
Comment (nerzhul) [inline src/content_mapblock.cpp]: no need for this you already have this in directiontables.h if i'm correct

**Comment by Wayward1** [comment]
Upon initial testing, FPS seems to be a bit higher, but movement is significantly more choppy, dtime jitter tends to skyrocket, and there are significant visual artifacts:

screenshot_20180110_143754
screenshot_20180110_143813
screenshot_20180110_143916

Note the dtime jitter here:
screenshot_20180110_144105

File: src/content_mapblock.cpp (hunk 1 above)
Comment (numberZero) [reply src/content_mapblock.cpp]: sadly no, the order is different, and it is important here.

**Comment by numberZero** [comment]
Fixed glitches. Grass glitches were due to the face grass side is not `tileable_vetical`; a nodebox would look the same.
screenshot_20180110_235121
But dtime jitter difference is sad.

**Comment by paramat** [comment]
Why drop 'fast faces' if that seems to be an optimisation?
What features do you have in mind to be based on this? It may not be worth doing this depending on the features you are considering.

**Comment by numberZero** [comment]
@paramat Depends on how fast these faces actually are. If they are significantly faster than this simple solution, it doesn’t worth that.

**Comment by Fixer-007** [comment]
I will look into it.

**Comment by Wayward1** [comment]
With further testing, FPS seems to be a bit lower, hovering around 52 versus the latest Minetest at around 57. The choppiness is much lower on this device, but still noticeable.

**Comment by Fixer-007** [comment]
Did 4 fly straight tests over mg7 terrain with PR and without (2 each), seems like minimal effect on my setup.
default
However, CPU usage seems higher, there were about 14-15% cpu cycles more.

// Core i3-2120 3,3GHz, ATI X800GTO 256mb OpenGL 2.1, win7

**Comment by rubenwardy** [comment]
What's the benefit and what PRs are relying on this?

**Comment by numberZero** [comment]
No PRs currently rely on this.
The benefit should be, less weird and fragile code amount. The benefit should be, reusing of same code where appropriate.
By the way, I found some interesting details that might be handy to fix by themselves, like nodeboxes not supporting non-tiling textures.

**Comment by Fixer-007** [comment]
>Only 15%?

Well, according to Process Explorer figures, maybe it is wrong to count CPU cycles? I don't use profilers or other stuff.

**Comment by lhofhansl** [comment]
So. Much. Simpler. :)  I definitely like that part of it.
Going to try it - although I also have a pretty fast machine.

**Comment by lhofhansl** [comment]
I see a measurable drop in FPS. Without the patch FPS is capped at 60 (have not tried with a higher limit), with the patch it hovers around 49.

If that's typical we should keep fast faces.

**Comment by numberZero** [comment]
@lhofhansl The reason for lower FPS may be lack of tiling, as the mesh my PR produce should be identical otherwise, and meshgen time shouldn’t affect FPS much. On my GPU, that does not affect FPS at all, vertex shaders take way less time than fragment ones. On your GPU the situation seems to be different.
Note: I suppose you tested in vanilla MTG (or `minimal`), full of flat solid areas. On messy landscape the FPS difference should be lower.
Nevertheless, this appeared to be a bad idea; fast faces should be improved and not dropped. Closing.
P.S. Thanks for testing.

**Comment by lhofhansl** [comment]
@numberZero Yeah, vanilla MTG. This was a "scene" with water, some islands, and some mountains visible.

**Comment by lhofhansl** [comment]
@numberZero I'm looking at this one again, as fast faces prevent many vertex shader effects, since you get lines between long faces.
I wonder if the rational for fastfaces still holds today. I think removing them would lead to more draw calls, right? That would be bad.

**Comment by numberZero** [comment]
> I think removing them would lead to more draw calls, right? That would be bad.

No-no, draw call number is not affected. Vertex number would be higher, sometimes (on plains) significantly.

**Comment by lhofhansl** [comment]
I see, not reducing the number of meshes, just the (average) count of vertices per mesh.

I'm going to try this again. The patch obviously does not apply anymore. I assume you did not keep a current version... (Had to ask, just in case.)

Tiling is already turned off for water (when waves are enabled).

**Comment by numberZero** [comment]
@lhofhansl to test rendering performance, just forcing `next_is_different` to `true` in the fastface generator should be enough.

**Comment by lhofhansl** [comment]
@numberZero I did that already :) I see the extra vertices rendered (F6 shows 8.8m instead of 6.9m), but FPS stays the same (31 FPS). This is a new laptop, so not sure whether that is typical. I'll play with it some more.

**Comment by numberZero** [comment]
Minetest is probably limited by drawcalls and not rendering itself on any more-or-less modern discrete GPU (except of maybe some nVidia® cards under nouveau which just can’t get them off the power-saving mode), and the former is mostly unaffected by fast-faces. But there are also integrated GPUs...

**Comment by lhofhansl** [comment]
And here I am looking at this again. This time for large view_ranges, where the fastface (getTileInfo) is *the* CPU hog.

**Pull request closed without merging**